### ISSUE TITLE
Meta information for data

### ISSUE THREAD

**Issue by dmpetrov** [iterative/dvc#8214]
We need to add meta-info to DVC files. Please use https://github.com/iterative/gto protect as an example of signals that are needed: description, labels, type, etc.

Out of the scope / Next steps: 
1. Commands to extract this metadata (initially).
2. Add custom meta field such as: custom metrics or summary statistics.
3. Metadata in pipeline.

**Comment by dberenbaum** [comment]
For now, let's say `type` and `label` are needed (and `description` if the existing field isn't enough).

**Comment by dberenbaum** [comment]
@dmpetrov For this issue, do we need the ability to add them from CLI? Like `dvc add --type`? Do they need to be shown in any existing commands yet?

**Comment by skshetry** [comment]
@dmpetrov, could you please elaborate on why we need this and how we plan to use this? In a way, we already have `meta` in `.dvc` and `dvc.yaml` files, which supports arbitrary information, but it's mostly intended for users.
But if it's to be consumed by external tools instead of DVC, it makes sense for those tools to also use `meta` fields.

Here, it seems like we need arbitrary information here too (for extensibility in the future), but it's not clear if `meta` field could be reused here.

**Comment by dmpetrov** [comment]
@skshetry yeah, the current meta might be enough.

Questions to clarify:
1. Can we add description, labels and type using meta?
2. Is it possible to add all 3 in a single command?
3. Is it possible to make it in `dvc add`?
4. (optional) Is there a command to show it?

**Comment by skshetry** [comment]
> 1. Can we add description, labels and type using meta?

yeah, the good thing about `meta` is that it is arbitrarily defined. It can have anything inside. 🙂 
The bad thing about `meta` is that it is arbitrarily defined. It can have anything inside. 

> 2. Is it possible to add all 3 in a single command?

We already have support for setting description in `import/import-url/add` with `--desc` flag.
We don't have support for setting arbitrary values in meta in commands yet, it is something that has to be set manually. But once set, `dvc` tries to preserve it even across overwrites/modifications.

> 3. Is it possible to make it in `dvc add`?
> > 4. (optional) Is there a command to show it?

yeah, it should be and no, there is no command to show it. This is because DVC considers `meta` purely for user needs, it does not try to interpret it in any way.

---

So my `meta` suggestion was because of the "future extensibility" requirements. I think if DVC is going to interpret them, it has to be structured, eg: `labels` have to be a list, `type` has to be of type string, etc. which currently won't work with `meta`, as it can have arbitrary information and is meant to be only used by users. So, if that's the case,
we'll need to add `labels` and `type` outside of `meta` (`desc` aka description is already supported). Something like follows:
```yaml
outs:
- md5: d3b07384d113edec49eaa6238ad5ff00
  size: 4
  path: model.pkl
  desc: My model
  type: model
  labels:
  - label
```

If DVC is not going to interpret, then `meta` could be reused.

**Comment by skshetry** [comment]
@dmpetrov @dberenbaum, I went ahead with `labels`/`type` field outside of `meta` as `desc` is already outside. 
The `add`/`import`/`import-url` commands all take `--desc`/`--type`/`--labels` flags.

> (optional) Is there a command to show it?

I am not sure what we want here, but I played with `dvc data ls` command and added some filtering mechanisms to it.

Please check out #8232 and we'll iterate from that.

**Comment by jorgeorpinel** [comment]
Hi

> > We need to add meta-info to DVC files
>
> The add/import/import-url commands all take --desc/--type/--labels flags

What about dvc.yaml `outs`? (`stage add -o`, maybe `repro`)

**Comment by skshetry** [comment]
> What about dvc.yaml `outs`? (`stage add -o`, `exp init`, maybe `repro/ exp run`)

Now `dvc.yaml` has `desc/label/type/meta` support inside `outs`. No support to add that with commands, and I am not sure if we need to tbh. In the future, there might be a separate command to modify these metadata, this is still experimental.

> And what happens to lock files?

No changes to lock files, as they `dvc.yaml` has these metadata.

**Comment by dberenbaum** [comment]
Closing as completed for now
